egs/wsj/s5/steps/libs/nnet3/xconfig/gru.py

```diff
@@ -0,0 +1,642 @@
+           ng_per_element_scale_options += " param-mean=0.0 param-stddev=1.0 "
+        pes_str = ng_per_element_scale_options
+
+        # formulation like:
+        # z_t = \sigmoid ( x_t * U^z + h_{t-1} * W^z ) // update gate
+        # r_t = \sigmoid ( x_t * U^r + h_{t-1} * W^r ) // reset gate
+        # \tilde{h}_t = \tanh ( x_t * U^h + ( h_{t-1} \dot r_t ) * W^h )
+        # h_t = ( 1 - z_t ) \dot \tilde{h}_t + z_t \dot h_{t-1}
+        # y_t = h_t // y_t is the output
+
+        # write bias and minus-scale
+        f = open(self.config['vars_path']+"/minus_one", 'w')
+        f.write(" [ ")
+        for i in range(cell_dim):
+            f.write("-1 ")
+        f.write("]\n")
+        f.close()
+
+        f = open(self.config['vars_path']+"/bias_one", 'w')
+        f.write(" [ ")
+        for i in range(cell_dim):
+            f.write("1 ")
+        f.write("]\n")
+        f.close()
+
+        configs = []
+        configs.append("# Update gate control : W_z* matrics")
+        configs.append("component name={0}.W_z.xs_z type=NaturalGradientAffineComponent input-dim={1} output-dim={2} {3}".format(name, input_dim + cell_dim, cell_dim, affine_str))
+        
+        configs.append("# Reset gate control : W_r* matrics")
+        configs.append("component name={0}.W_z.xs_r type=NaturalGradientAffineComponent input-dim={1} output-dim={2} {3}".format(name, input_dim + cell_dim, cell_dim, affine_str))
+
+        configs.append("# h related matrix : W_h* matrics")
+        configs.append("component name={0}.W_h.UW type=NaturalGradientAffineComponent input-dim={1} output-dim={2} {3}".format(name, input_dim + cell_dim, cell_dim , affine_str))
+        
+        configs.append("# Defining the non-linearities")
+        configs.append("component name={0}.z type=SigmoidComponent dim={1} {2}".format(name, cell_dim, repair_nonlin_str))
+        configs.append("component name={0}.r type=SigmoidComponent dim={1} {2}".format(name, cell_dim, repair_nonlin_str))
+        configs.append("component name={0}.h type=TanhComponent dim={1} {2}".format(name, cell_dim, repair_nonlin_str))
+
+        configs.append("# Defining the components for other cell computations")
+        configs.append("component name={0}.h1 type=ElementwiseProductComponent input-dim={1} output-dim={2}".format(name, 2 * cell_dim, cell_dim))
+        configs.append("component name={0}.y1 type=ElementwiseProductComponent input-dim={1} output-dim={2}".format(name, 2 * cell_dim, cell_dim))
+        configs.append("component name={0}.y2 type=ElementwiseProductComponent input-dim={1} output-dim={2}".format(name, 2 * cell_dim, cell_dim))
+        configs.append("component name={0}.y type=NoOpComponent dim={1}".format(name, cell_dim))
+
+        configs.append("# Defining fixed scale/bias component for (1 - z_t)")
+        configs.append("component name={0}.fixed_scale_minus_one type=FixedScaleComponent scales={1}".format(name, self.config['vars_path']+"/minus_one"))
```

Comment:
there is no need for 'vars_path' here.  FixedScaleComponent defaults to a scale of 1.0, you just have to supply the dim; and FixedBiasComponent can have a constant bias specified by "bias=1.0 dim=xxx".

As soon as I merge this to the RNNLM branch, though, I'll ask you to simplify these config-generation scripts to use the offset and scale capabilities I added there.